src/freenet/client/filter/GIFFilter.java

```diff
@@ -30,22 +35,452 @@
+					default:
+						// The specification expects us to skip other data; we can simply omit it.
+				}
+			}
+			if (!imageSeen) {
+				throwDataError(l10n("noDataTitle"), l10n("noData"));
+			}
+			if (!terminated) {
+				throwDataError(l10n("unterminatedGifTitle"), l10n("unterminatedGif"));
+			}
+			// There may still be trailing data at this point; we can simply omit it.
+			writeByte(GIF_TERMINATOR);
+		}
+
+		/** Filters a render block. Actual LZW data is *not* checked. */
+		private boolean filterImage() throws IOException, DataFilterException {
+			final int imageLeft = readShort();
+			final int imageTop = readShort();
+			final int imageWidth = readShort();
+			final int imageHeight = readShort();
+			final int imageFlags = readByte();
+			final boolean hasLocalColorMap = (imageFlags & 0x80) == 0x80;
+			final byte[] localColorMap;
+			if (hasLocalColorMap) {
+				final int bitsPerPixel = (imageFlags & 0x07) + 1;
+				final int imageColors = 1 << bitsPerPixel;
+				localColorMap = readBytes(3 * imageColors);
+			} else {
+				localColorMap = new byte[0];
+			}
+			final int lzwCodeSize = readByte();
+			if (imageLeft + imageWidth > screenWidth || imageTop + imageHeight > screenHeight ||
+					lzwCodeSize < 2 || lzwCodeSize >= 12 || !validateImageFlags(imageFlags)) {
+				foundImageData(false);
+				skipSubBlocks();
+				return false;
+			} else {
+				foundImageData(true);
+				writeByte(0x2c);
+				writeShort(imageLeft);
+				writeShort(imageTop);
+				writeShort(imageWidth);
+				writeShort(imageHeight);
+				writeByte(imageFlags);
+				writeBytes(localColorMap);
+				writeByte(lzwCodeSize);
+				copySubBlocks();
+				return true;
+			}
+		}
+
+		/** Skips an entire extension block; assumes the extension indicator is already read. */
+		private void skipExtensionBlock() throws IOException {
+			skip(1); // extension function
+			skipSubBlocks();
+		}
+
+		/** Skips all subblocks in the input, until the empty terminator subblock is found. */
+		protected final void skipSubBlocks() throws IOException {
+			int length;
+			while ((length = readByte()) != 0) {
+				skip(length);
+			}
+		}
+
+		/** Copies all subblocks to the output, until the empty terminator subblock is found. */
+		protected final void copySubBlocks() throws IOException {
+			int length;
+			while ((length = readByte()) != 0) {
+				writeByte(length);
+				copy(length);
+			}
+			writeByte(0);
+		}
+
+		/** Copy a small number of bytes from input to output. */
+		protected final void copy(int length) throws IOException {
+			for (int i = 0; i < length; i++) {
+				writeByte(readByte());
+			}
+		}
+
+		/** Read an unsigned byte from the input. */
+		protected final int readByte() throws IOException {
+			int val = input.read();
+			if (val == -1) {
+				throw new EOFException();
+			}
+			return val;
+		}
+
+		/** Write an unsigned byte to the output. */
+		protected final void writeByte(int val) throws IOException {
+			assert(val >= 0x00 && val <= 0xFF);
```

Comment:
The assertions here are mostly to make the intention of the code clear: to human readers and to static code analysis tools. If only Java would have unsigned bytes and shorts… neither this or excessive casting (and `& 0xFF` tricks) would be required :)

Would you rather have it throw `IllegalArgumentException` or `AssertionError`, or just assume we are doing the right thing all along?